### PR TITLE
Make dynamo wrapped modules work with save_pretrained 

### PR DESCRIPTION
**Pull request by pcuenca**
Experimental, not sure if there's a better way to do it. We'd also need to do something similar in `modeling_utils.py`.

Addresses #2709.

### CODE REVIEW

**Comment by sayakpaul** [comment]
I don't see other way around either.

File: src/diffusers/pipelines/pipeline_utils.py [hunk 1]
```diff
@@ -484,6 +484,11 @@ def is_saveable_module(name, value):
             sub_model = getattr(self, pipeline_component_name)
             model_cls = sub_model.__class__
 
+            # Dynamo wraps the original mode and changes the class.
+            # Is there a principled way to obtain the original class?
+            if "_dynamo.eval_frame" in str(model_cls):
+                model_cls = sub_model._orig_mod.__class__
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/pipeline_utils.py:490]:
```suggestion
                model_cls = getattr(self.pipe.config, pipeline_component_name)[-1]
```

File: src/diffusers/pipelines/pipeline_utils.py [hunk 2]
```diff
@@ -484,6 +484,11 @@ def is_saveable_module(name, value):
             sub_model = getattr(self, pipeline_component_name)
             model_cls = sub_model.__class__
 
+            # Dynamo wraps the original mode and changes the class.
+            # Is there a principled way to obtain the original class?
+            if "_dynamo.eval_frame" in str(model_cls):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/pipeline_utils.py:489]: Think that's a bit cleaner
```suggestion
            if is_torch_version(">=", "2.0.0") and isinstance(sub_model, torch._dynamo.eval_frame.OptimizedModule):
```

File: src/diffusers/pipelines/pipeline_utils.py (hunk 1 above)
Comment (patrickvonplaten) [inline src/diffusers/pipelines/pipeline_utils.py:490]: Actually think that's quite clean - another option would be the following:
```suggestion
                model_cls = sub_model.config._class_name
```
But think what you had before is actually better (maybe with a small comment)

**Review by patrickvonplaten — COMMENTED** [review]
Looks nice - thanks for the fix! Let's maybe do the same in `modeling_utils.py`. Would just recommend to use the `isinstance(...)` check instead of string matching check - think that's a bit less brittle. 

The way you retrieve the original class here is good IMO

**Comment by pcuenca** [comment]
I created a simple test that unsurfaced additional problems in the configuration (the wrong class was saved to config.json).

**Review by sayakpaul — APPROVED** [review]
Thanks!

File: src/diffusers/pipelines/pipeline_utils.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/pipeline_utils.py:500]: @pcuenca quick question - why don't we go for this API? Think `isinstance` is a better check here than the string checking

File: src/diffusers/pipelines/pipeline_utils.py [hunk 3]
```diff
@@ -419,6 +426,10 @@ def register_modules(self, **kwargs):
             if module is None:
                 register_dict = {name: (None, None)}
             else:
+                # register the original module, not the dynamo compiled one
+                if "_dynamo.eval_frame" in str(module.__class__):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/pipeline_utils.py:430]: Think this check is better:
```py
if is_torch_version(">=", "2.0.0") and isinstance(sub_model, torch._dynamo.eval_frame.OptimizedModule):
```

File: src/diffusers/pipelines/pipeline_utils.py [hunk 4]
```diff
@@ -255,7 +255,14 @@ def maybe_raise_or_warn(
             if class_candidate is not None and issubclass(class_obj, class_candidate):
                 expected_class_obj = class_candidate
 
-        if not issubclass(passed_class_obj[name].__class__, expected_class_obj):
+        # Dynamo wraps the original model in a private class.
+        # I didn't find a public API to get the original class.
+        sub_model = passed_class_obj[name]
+        model_cls = sub_model.__class__
+        if "_dynamo.eval_frame" in str(model_cls):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/pipeline_utils.py:262]: Think this check is better:
```py
if is_torch_version(">=", "2.0.0") and isinstance(sub_model, torch._dynamo.eval_frame.OptimizedModule):
```

**Review by patrickvonplaten — COMMENTED** [review]
Can we apply the same changes to the model save_pretrained function?

File: src/diffusers/pipelines/pipeline_utils.py (hunk 2 above)
Comment (pcuenca) [reply src/diffusers/pipelines/pipeline_utils.py:500]: Because `torch._dynamo` is not available in all platforms, just on cuda builds.

File: src/diffusers/pipelines/pipeline_utils.py (hunk 3 above)
Comment (pcuenca) [reply src/diffusers/pipelines/pipeline_utils.py:430]: As commented above, `torch._dynamo` does not exist in all platforms.

Comment (patrickvonplaten) [reply src/diffusers/pipelines/pipeline_utils.py:430]: Ah sorry, now I get it

File: src/diffusers/pipelines/pipeline_utils.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/pipeline_utils.py:500]: Think that's a bit cleaner
```suggestion
            if is_torch_version(">=", "2.0.0") and hasattr(torch, "_dynamo") and isinstance(sub_model, torch._dynamo.eval_frame.OptimizedModule):
```

**Comment by patrickvonplaten** [comment]
Thanks for clarifying the design here. Could we then maybe go with: https://github.com/huggingface/diffusers/pull/2726#discussion_r1143232060 or do you prefer how it is now?

**Comment by pcuenca** [comment]
> Can we apply the same changes to the model save_pretrained function?

It works, actually; at first I thought it wouldn't but I was wrong. When `save_pretrained` is invoked on a compiled model, `self` is the appropriate class and saving succeeds. I verified that the json file is correct and it refers to the right class. I also added a very small test case.

**Comment by pcuenca** [comment]
@patrickvonplaten so the only thing missing here would be to decide whether to use a helper function to verify the class. I'm undecided because I think that testing for a private class that might change is not very much better than testing for a name, but we might want to make it fail so we can fix it when it changes. What do you think?

File: src/diffusers/pipelines/pipeline_utils.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/pipeline_utils.py:500]: I see! Tbh, both ways are fine by me - I do prefer to avoid silent errors and generally don't like string "in" testing for classes as this is a very brittle test in my opinion. 

Would prefer a helper function here, but also ok to leave as is as it's an edge case.

**Review by patrickvonplaten — APPROVED** [review]
Thanks! Summarized my thoughts here: https://github.com/huggingface/diffusers/pull/2726/files#r1146130475 

Good to go whenever you want :-)

**Comment by sayakpaul** [comment]
Nice! Looks clean to me :)
